nab/sweeper.py

```diff
@@ -0,0 +1,245 @@
+
+
+
+class Sweeper(object):
+
+
+  def __init__(self, probationPercent=0.15, costMatrix=None):
+    self.probationPercent = probationPercent
+
+    self.tpWeight = None
+    self.fpWeight = None
+    self.fnWeight = None
+
+    if costMatrix is not None:
+      self.setCostMatrix(costMatrix)
+
+
+  def setCostMatrix(self, costMatrix):
+    self.tpWeight = costMatrix["tpWeight"]
+    self.fpWeight = costMatrix["fpWeight"]
+    self.fnWeight = costMatrix["fnWeight"]
+
+
+  def _getProbationaryLength(self, numRows):
+    return min(math.floor(self.probationPercent * numRows), self.probationPercent * 5000)
+
+
+  def _prepAnomalyListForScoring(self, inputAnomalyList):
+    """Sort by anomaly score and filter all rows with 'probationary' window name"""
+    return sorted(
+      [x for x in inputAnomalyList if x.windowName != 'probationary'],
+      key=lambda x: x.anomalyScore,
+      reverse=True)
+
+
+  def _prepareScoreByThresholdParts(self, inputAnomalyList):
+    scoreParts = {"fp": 0}
+    for row in inputAnomalyList:
+      if row.windowName not in ('probationary', None):
+        scoreParts[row.windowName] = -self.fnWeight
+    return scoreParts
+
+
+  def calcSweepScore(self, timestamps, anomalyScores, windowLimits, dataSetName):
+    assert len(timestamps) == len(anomalyScores), "timestamps and anomalyScores should not be different lengths!"
+    timestamps = list(timestamps)
+    windowLimits = list(windowLimits)  # Copy because we mutate this list
+    # The final list of anomaly points returned from this function.
+    # Used for threshold optimization and scoring in other functions.
+    anomalyList = []
+
+    # One-time config variables
+    maxTP = scaledSigmoid(-1.0)
+    probationaryLength = self._getProbationaryLength(len(timestamps))
+
+    # Iteration variables - these update as we iterate through the data
+    curWindowLimits = None
+    curWindowName = None
+    curWindowWidth = None
+    curWindowRightIndex = None
+    prevWindowWidth = None
+    prevWindowRightIndex = None
+
+    for i, (curTimestamp, curAnomaly) in enumerate(zip(timestamps, anomalyScores)):
+      unweightedScore = None
+      weightedScore = None
+
+      # If not in a window, check if we've just entered one
+      if len(windowLimits) > 0 and curTimestamp == windowLimits[0][0]:
+        curWindowLimits = windowLimits.pop(0)
+        curWindowName = "%s|%s" % (dataSetName, curWindowLimits[0])
+        curWindowRightIndex = timestamps.index(curWindowLimits[1])
+        curWindowWidth = float(curWindowRightIndex - timestamps.index(curWindowLimits[0]) + 1)
+
+        logger.debug("Entering window: %s (%s)" % (curWindowName, str(curWindowLimits)))
+
+      # If in a window, score as if true positive
+      if curWindowLimits is not None:
+        # Doesn't the `+ 1` in this equation mean we can _never_ have a positionInWindow == 0?
+        positionInWindow = -(curWindowRightIndex - i + 1) / curWindowWidth
```

Comment:
See [above](https://github.com/numenta/NAB/pull/328/files#diff-6956a7129fa553c42f4a85415f85924fR137):

```python
curWindowRightIndex = timestamps.index(curWindowLimits[1])
```
Isn't that where the `1` is coming from?

Reply:
Yeah, makes sense, I think.  I removed the question/comment.